posts/unified-memory-oversubscription/uvm_oversubs.cu

```diff
@@ -0,0 +1,501 @@
+    access_str = "streaming,";
+  else if (memory_access == BLOCK_STREAMING)
+    access_str = "block_streaming,";
+  else if (memory_access == RANDOM_WARP)
+    access_str = "random_warp,";  
+  header_string += access_str;
+  header_string += std::to_string(oversubscription_factor);
+  header_string += ",";
+  
+  if (page_size == TWO_MB)
+    header_string += "2MB,";
+  else if (page_size == SIXTY_FOUR_KB)
+    header_string += "64KB,";
+  else if (page_size == FOUR_KB)
+  header_string += "4KB,";
+
+  header_string += "blocksize=";
+  header_string += std::to_string(block_size);
+  header_string += ",";
+
+
+  header_string += "loop_count=";
+  header_string += std::to_string(loop_count);
+
+  // determine cudaMallocManaged size
+  int current_device = 0;
+  CUDA_CHECK(cudaSetDevice(current_device));
+  cudaDeviceProp prop;
+  CUDA_CHECK(cudaGetDeviceProperties(&prop, current_device));
+  bool is_P9 = (prop.pageableMemoryAccessUsesHostPageTables == 1);
+
+  size_t allocation_size = (size_t)(oversubscription_factor * prop.totalGlobalMem);
+
+  void *blockMemory = nullptr;
+  if (memory_access != STREAMING && uvm_behavior) {
+    // reduce test working memory
+    // cudaMalloc 2/3 GPU
+    size_t cudaMallocSize = AlignSize(size_t(prop.totalGlobalMem * 0.67), page_size);
+    allocation_size = AlignSize(size_t(prop.totalGlobalMem * 0.33 * oversubscription_factor),
+                                  page_size);
+    CUDA_CHECK(cudaMalloc(&blockMemory, cudaMallocSize));
+  }
+  // pad allocation to page_size
+  allocation_size = AlignSize(allocation_size, page_size);
+  size_t num_pages = allocation_size / page_size;
+
+  size_t avail_phy_vidmem = 0, total_phy_vidmem = 0;
+  // allocate memory - add hints etc as needed
+  void *uvm_alloc_ptr = NULL;
+
+  // For P9 we need to allocate and free in-benchmark loop
+  // as evicted memory has remote mappings don't trigger a page-fault
+  if (!(is_P9 && uvm_behavior == PAGE_FAULT)) {
+    CUDA_CHECK(cudaMallocManaged(&uvm_alloc_ptr, allocation_size));
+    CUDA_CHECK(cudaMemGetInfo(&avail_phy_vidmem, &total_phy_vidmem));
+    // populate pages on GPU
+    CUDA_CHECK(cudaMemPrefetchAsync(uvm_alloc_ptr, allocation_size, current_device));
+  }
+
+  // P9 need more state space on vidmem - size in MB
+  size_t state_space_size = (prop.pageableMemoryAccessUsesHostPageTables == 1) ? 320 : 128;
+  size_t permissible_phys_pages_count = avail_phy_vidmem / page_size;
+  permissible_phys_pages_count -= (state_space_size * 1024 * 1024 / page_size);
+
+  dim3 block(block_size,1,1);
+  dim3 grid((prop.multiProcessorCount * prop.maxThreadsPerMultiProcessor) / block.x, 1, 1);
+  int num_blocks_per_sm = 1;  // placeholder value
+
+  cudaStream_t task_stream;
+  CUDA_CHECK(cudaStreamCreate(&task_stream));
+
+  cudaEvent_t startE, stopE;
+  CUDA_CHECK(cudaEventCreate(&startE));
+  CUDA_CHECK(cudaEventCreate(&stopE));
+
+  float kernel_time = 0.0f;
+  float accum_kernel_time = 0.0f;
+  float accum_bw = 0.0f;
```

Comment:
The preparation/setup code is quite long, is there a way we can move it to a separate function to have main a bit easier to read. I think the main piece of code that people would be interested in is the below for-loop that we measure with data placement / hints + calling the kernel, and it takes a while to get to it.

Reply:
Refactored the main function to be more concise.